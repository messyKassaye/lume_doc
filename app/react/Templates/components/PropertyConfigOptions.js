import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { FilterSuggestions } from 'app/Templates/components/FilterSuggestions';
import { connect } from 'react-redux';

import PropertyConfigOption from './PropertyConfigOption';
import Tip from '../../Layout/Tip';

class PropertyConfigOptions extends Component {
  render() {
    const { index, filter, type, canBeFilter } = this.props;
    return (
      <div>
        <PropertyConfigOption
          label="Hide label"
          model={`template.data.properties[${index}].noLabel`}
        >
          <Tip>This property will be shown without the label.</Tip>
        </PropertyConfigOption>
        <PropertyConfigOption
          label="Required property"
          model={`template.data.properties[${index}].required`}
        >
          <Tip>You won&#39;t be able to publish a document if this property is empty.</Tip>
        </PropertyConfigOption>
        <PropertyConfigOption
          label="Show in cards"
          model={`template.data.properties[${index}].showInCard`}
        >
          <Tip>This property will appear in the library cards as part of the basic info.</Tip>
        </PropertyConfigOption>

        {canBeFilter && (
          <div className="inline-group">
            <PropertyConfigOption
              label="Use as filter"
              model={`template.data.properties[${index}].filter`}
            >
              <Tip>
                This property will be used for filtering the library results. When properties match
                in equal name and field type with other document types, they will be combined for
                filtering.
              </Tip>
            </PropertyConfigOption>
            {filter && (
              <>
                <PropertyConfigOption
                  label="Default filter"
                  model={`template.data.properties[${index}].defaultfilter`}
                >
                  <Tip>
                    Use this property as a default filter in the library. When there are no document
                    types selected, this property will show as a default filter for your collection.
                  </Tip>
                </PropertyConfigOption>
                {['text', 'date', 'numeric', 'select'].includes(type) && (
                  <PropertyConfigOption
                    label="Priority sorting"
                    model={`template.data.properties[${index}].prioritySorting`}
                  >
                    <Tip>
                      Properties marked as priority sorting will be used as default sorting
                      criteria. If more than one property is marked as priority sorting the system
                      will try to pick-up the best fit. When listing mixed template types, the
                      system will pick-up the best combined priority sorting.
                    </Tip>
                  </PropertyConfigOption>
                )}
              </>
            )}
          </div>
        )}
        <FilterSuggestions index={index} />
      </div>
    );
  }
}

PropertyConfigOptions.defaultProps = {
  canBeFilter: true,
  filter: false,
};

PropertyConfigOptions.propTypes = {
  filter: PropTypes.bool,
  index: PropTypes.number.isRequired,
  type: PropTypes.string.isRequired,
  canBeFilter: PropTypes.bool,
};

export function mapStateToProps({ template }, props) {
  return {
    filter: template.data.properties[props.index].filter,
  };
}

export default connect(mapStateToProps)(PropertyConfigOptions);
