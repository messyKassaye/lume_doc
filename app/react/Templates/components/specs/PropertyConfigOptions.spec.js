import PropertyConfigOptions from '../PropertyConfigOptions';
import { renderConnected } from '../../specs/utils/renderConnected.tsx';

describe('PropertyConfigOptions', () => {
  let component;
  let props;
  let storeData;

  beforeEach(() => {
    props = {
      index: 2,
      type: 'select',
      property: {},
    };

    storeData = {
      template: {
        data: {
          properties: [
            { name: 'title', label: 'Title' },
            { name: 'creationDate' },
            { name: 'property2' },
          ],
        },
      },
    };
  });

  const expectMatch = () => {
    component = renderConnected(PropertyConfigOptions, props, storeData);
    expect(component).toMatchSnapshot();
  };

  it('should render fields with the correct datas', () => {
    expectMatch();
  });

  describe('Once the property is checked as filter', () => {
    it('should render the default filter option', () => {
      storeData.template.data.properties[2].filter = true;
      expectMatch();
    });
  });

  describe('priority sorting option', () => {
    describe('when property filter is true', () => {
      it('should render for text, date, numeric and select if property filter is true', () => {
        storeData.template.data.properties[2].filter = true;
        props.type = 'text';
        expectMatch();
        props.type = 'date';
        expectMatch();
        props.type = 'numeric';
        expectMatch();
        props.type = 'select';
        expectMatch();
      });
    });
    describe('when property filter is not true', () => {
      it('should not render priority sorting option', () => {
        storeData.template.data.properties[2].filter = false;
        props.type = 'text';
        expectMatch();
        props.type = 'date';
        expectMatch();
        props.type = 'numeric';
        expectMatch();
        props.type = 'select';
        expectMatch();
      });
    });
  });

  describe('Additional options', () => {
    it('should allow to exclude the "use as filter" option', () => {
      props.canBeFilter = false;
      expectMatch();
    });
  });
});
