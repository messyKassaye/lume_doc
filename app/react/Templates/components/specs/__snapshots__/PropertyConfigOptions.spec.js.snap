// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyConfigOptions Additional options should allow to exclude the "use as filter" option 1`] = `
<div>
  <PropertyConfigOption
    label="Hide label"
    model="template.data.properties[2].noLabel"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will be shown without the label.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Required property"
    model="template.data.properties[2].required"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      You won't be able to publish a document if this property is empty.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Show in cards"
    model="template.data.properties[2].showInCard"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will appear in the library cards as part of the basic info.
    </Tip>
  </PropertyConfigOption>
  <Connect(SharedProperties)
    index={2}
  />
</div>
`;

exports[`PropertyConfigOptions Once the property is checked as filter should render the default filter option 1`] = `
<div>
  <PropertyConfigOption
    label="Hide label"
    model="template.data.properties[2].noLabel"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will be shown without the label.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Required property"
    model="template.data.properties[2].required"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      You won't be able to publish a document if this property is empty.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Show in cards"
    model="template.data.properties[2].showInCard"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will appear in the library cards as part of the basic info.
    </Tip>
  </PropertyConfigOption>
  <div
    className="inline-group"
  >
    <PropertyConfigOption
      label="Use as filter"
      model="template.data.properties[2].filter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        This property will be used for filtering the library results. When properties match in equal name and field type with other document types, they will be combined for filtering.
      </Tip>
    </PropertyConfigOption>
    <PropertyConfigOption
      label="Default filter"
      model="template.data.properties[2].defaultfilter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        Use this property as a default filter in the library. When there are no document types selected, this property will show as a default filter for your collection.
      </Tip>
    </PropertyConfigOption>
    <PropertyConfigOption
      label="Priority sorting"
      model="template.data.properties[2].prioritySorting"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        Properties marked as priority sorting will be used as default sorting criteria. If more than one property is marked as priority sorting the system will try to pick-up the best fit. When listing mixed template types, the system will pick-up the best combined priority sorting.
      </Tip>
    </PropertyConfigOption>
  </div>
  <Connect(SharedProperties)
    index={2}
  />
</div>
`;

exports[`PropertyConfigOptions priority sorting option when property filter is not true should not render priority sorting option 1`] = `
<div>
  <PropertyConfigOption
    label="Hide label"
    model="template.data.properties[2].noLabel"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will be shown without the label.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Required property"
    model="template.data.properties[2].required"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      You won't be able to publish a document if this property is empty.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Show in cards"
    model="template.data.properties[2].showInCard"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will appear in the library cards as part of the basic info.
    </Tip>
  </PropertyConfigOption>
  <div
    className="inline-group"
  >
    <PropertyConfigOption
      label="Use as filter"
      model="template.data.properties[2].filter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        This property will be used for filtering the library results. When properties match in equal name and field type with other document types, they will be combined for filtering.
      </Tip>
    </PropertyConfigOption>
  </div>
  <Connect(SharedProperties)
    index={2}
  />
</div>
`;

exports[`PropertyConfigOptions priority sorting option when property filter is not true should not render priority sorting option 2`] = `
<div>
  <PropertyConfigOption
    label="Hide label"
    model="template.data.properties[2].noLabel"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will be shown without the label.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Required property"
    model="template.data.properties[2].required"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      You won't be able to publish a document if this property is empty.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Show in cards"
    model="template.data.properties[2].showInCard"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will appear in the library cards as part of the basic info.
    </Tip>
  </PropertyConfigOption>
  <div
    className="inline-group"
  >
    <PropertyConfigOption
      label="Use as filter"
      model="template.data.properties[2].filter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        This property will be used for filtering the library results. When properties match in equal name and field type with other document types, they will be combined for filtering.
      </Tip>
    </PropertyConfigOption>
  </div>
  <Connect(SharedProperties)
    index={2}
  />
</div>
`;

exports[`PropertyConfigOptions priority sorting option when property filter is not true should not render priority sorting option 3`] = `
<div>
  <PropertyConfigOption
    label="Hide label"
    model="template.data.properties[2].noLabel"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will be shown without the label.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Required property"
    model="template.data.properties[2].required"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      You won't be able to publish a document if this property is empty.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Show in cards"
    model="template.data.properties[2].showInCard"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will appear in the library cards as part of the basic info.
    </Tip>
  </PropertyConfigOption>
  <div
    className="inline-group"
  >
    <PropertyConfigOption
      label="Use as filter"
      model="template.data.properties[2].filter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        This property will be used for filtering the library results. When properties match in equal name and field type with other document types, they will be combined for filtering.
      </Tip>
    </PropertyConfigOption>
  </div>
  <Connect(SharedProperties)
    index={2}
  />
</div>
`;

exports[`PropertyConfigOptions priority sorting option when property filter is not true should not render priority sorting option 4`] = `
<div>
  <PropertyConfigOption
    label="Hide label"
    model="template.data.properties[2].noLabel"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will be shown without the label.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Required property"
    model="template.data.properties[2].required"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      You won't be able to publish a document if this property is empty.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Show in cards"
    model="template.data.properties[2].showInCard"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will appear in the library cards as part of the basic info.
    </Tip>
  </PropertyConfigOption>
  <div
    className="inline-group"
  >
    <PropertyConfigOption
      label="Use as filter"
      model="template.data.properties[2].filter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        This property will be used for filtering the library results. When properties match in equal name and field type with other document types, they will be combined for filtering.
      </Tip>
    </PropertyConfigOption>
  </div>
  <Connect(SharedProperties)
    index={2}
  />
</div>
`;

exports[`PropertyConfigOptions priority sorting option when property filter is true should render for text, date, numeric and select if property filter is true 1`] = `
<div>
  <PropertyConfigOption
    label="Hide label"
    model="template.data.properties[2].noLabel"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will be shown without the label.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Required property"
    model="template.data.properties[2].required"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      You won't be able to publish a document if this property is empty.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Show in cards"
    model="template.data.properties[2].showInCard"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will appear in the library cards as part of the basic info.
    </Tip>
  </PropertyConfigOption>
  <div
    className="inline-group"
  >
    <PropertyConfigOption
      label="Use as filter"
      model="template.data.properties[2].filter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        This property will be used for filtering the library results. When properties match in equal name and field type with other document types, they will be combined for filtering.
      </Tip>
    </PropertyConfigOption>
    <PropertyConfigOption
      label="Default filter"
      model="template.data.properties[2].defaultfilter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        Use this property as a default filter in the library. When there are no document types selected, this property will show as a default filter for your collection.
      </Tip>
    </PropertyConfigOption>
    <PropertyConfigOption
      label="Priority sorting"
      model="template.data.properties[2].prioritySorting"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        Properties marked as priority sorting will be used as default sorting criteria. If more than one property is marked as priority sorting the system will try to pick-up the best fit. When listing mixed template types, the system will pick-up the best combined priority sorting.
      </Tip>
    </PropertyConfigOption>
  </div>
  <Connect(SharedProperties)
    index={2}
  />
</div>
`;

exports[`PropertyConfigOptions priority sorting option when property filter is true should render for text, date, numeric and select if property filter is true 2`] = `
<div>
  <PropertyConfigOption
    label="Hide label"
    model="template.data.properties[2].noLabel"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will be shown without the label.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Required property"
    model="template.data.properties[2].required"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      You won't be able to publish a document if this property is empty.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Show in cards"
    model="template.data.properties[2].showInCard"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will appear in the library cards as part of the basic info.
    </Tip>
  </PropertyConfigOption>
  <div
    className="inline-group"
  >
    <PropertyConfigOption
      label="Use as filter"
      model="template.data.properties[2].filter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        This property will be used for filtering the library results. When properties match in equal name and field type with other document types, they will be combined for filtering.
      </Tip>
    </PropertyConfigOption>
    <PropertyConfigOption
      label="Default filter"
      model="template.data.properties[2].defaultfilter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        Use this property as a default filter in the library. When there are no document types selected, this property will show as a default filter for your collection.
      </Tip>
    </PropertyConfigOption>
    <PropertyConfigOption
      label="Priority sorting"
      model="template.data.properties[2].prioritySorting"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        Properties marked as priority sorting will be used as default sorting criteria. If more than one property is marked as priority sorting the system will try to pick-up the best fit. When listing mixed template types, the system will pick-up the best combined priority sorting.
      </Tip>
    </PropertyConfigOption>
  </div>
  <Connect(SharedProperties)
    index={2}
  />
</div>
`;

exports[`PropertyConfigOptions priority sorting option when property filter is true should render for text, date, numeric and select if property filter is true 3`] = `
<div>
  <PropertyConfigOption
    label="Hide label"
    model="template.data.properties[2].noLabel"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will be shown without the label.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Required property"
    model="template.data.properties[2].required"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      You won't be able to publish a document if this property is empty.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Show in cards"
    model="template.data.properties[2].showInCard"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will appear in the library cards as part of the basic info.
    </Tip>
  </PropertyConfigOption>
  <div
    className="inline-group"
  >
    <PropertyConfigOption
      label="Use as filter"
      model="template.data.properties[2].filter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        This property will be used for filtering the library results. When properties match in equal name and field type with other document types, they will be combined for filtering.
      </Tip>
    </PropertyConfigOption>
    <PropertyConfigOption
      label="Default filter"
      model="template.data.properties[2].defaultfilter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        Use this property as a default filter in the library. When there are no document types selected, this property will show as a default filter for your collection.
      </Tip>
    </PropertyConfigOption>
    <PropertyConfigOption
      label="Priority sorting"
      model="template.data.properties[2].prioritySorting"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        Properties marked as priority sorting will be used as default sorting criteria. If more than one property is marked as priority sorting the system will try to pick-up the best fit. When listing mixed template types, the system will pick-up the best combined priority sorting.
      </Tip>
    </PropertyConfigOption>
  </div>
  <Connect(SharedProperties)
    index={2}
  />
</div>
`;

exports[`PropertyConfigOptions priority sorting option when property filter is true should render for text, date, numeric and select if property filter is true 4`] = `
<div>
  <PropertyConfigOption
    label="Hide label"
    model="template.data.properties[2].noLabel"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will be shown without the label.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Required property"
    model="template.data.properties[2].required"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      You won't be able to publish a document if this property is empty.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Show in cards"
    model="template.data.properties[2].showInCard"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will appear in the library cards as part of the basic info.
    </Tip>
  </PropertyConfigOption>
  <div
    className="inline-group"
  >
    <PropertyConfigOption
      label="Use as filter"
      model="template.data.properties[2].filter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        This property will be used for filtering the library results. When properties match in equal name and field type with other document types, they will be combined for filtering.
      </Tip>
    </PropertyConfigOption>
    <PropertyConfigOption
      label="Default filter"
      model="template.data.properties[2].defaultfilter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        Use this property as a default filter in the library. When there are no document types selected, this property will show as a default filter for your collection.
      </Tip>
    </PropertyConfigOption>
    <PropertyConfigOption
      label="Priority sorting"
      model="template.data.properties[2].prioritySorting"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        Properties marked as priority sorting will be used as default sorting criteria. If more than one property is marked as priority sorting the system will try to pick-up the best fit. When listing mixed template types, the system will pick-up the best combined priority sorting.
      </Tip>
    </PropertyConfigOption>
  </div>
  <Connect(SharedProperties)
    index={2}
  />
</div>
`;

exports[`PropertyConfigOptions should render fields with the correct datas 1`] = `
<div>
  <PropertyConfigOption
    label="Hide label"
    model="template.data.properties[2].noLabel"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will be shown without the label.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Required property"
    model="template.data.properties[2].required"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      You won't be able to publish a document if this property is empty.
    </Tip>
  </PropertyConfigOption>
  <PropertyConfigOption
    label="Show in cards"
    model="template.data.properties[2].showInCard"
  >
    <Tip
      icon="question-circle"
      position=""
    >
      This property will appear in the library cards as part of the basic info.
    </Tip>
  </PropertyConfigOption>
  <div
    className="inline-group"
  >
    <PropertyConfigOption
      label="Use as filter"
      model="template.data.properties[2].filter"
    >
      <Tip
        icon="question-circle"
        position=""
      >
        This property will be used for filtering the library results. When properties match in equal name and field type with other document types, they will be combined for filtering.
      </Tip>
    </PropertyConfigOption>
  </div>
  <Connect(SharedProperties)
    index={2}
  />
</div>
`;
